(DOCSP-45012) Transactions (#71)
mongodb/docs-cpp: source/write/transactions.txt

.. _cpp-transactions:

============
Transactions
============

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, rollback, undo operation

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use the {+driver-long+} to perform
**transactions**. Transactions allow
you to run a series of operations that do not change any data until the
transaction is committed. If any operation in the transaction returns an
error, the driver cancels the transaction and discards all data changes
before they ever become visible.

In MongoDB, transactions run within logical **sessions**. A
session is a grouping of related read or write operations that you intend to run sequentially. 

> Sessions enable causal consistency for a group of operations in an  **ACID-compliant**  transaction, which is a 

Reviewer: For causal consistency, I do think it can be an unfamiliar term so I think it would be good to provide some sort of reference (doesn't have to be a link - could be a short definition). I don't feel too strongly about it though, so I defer to your judgement!

Before you begin a transaction, you must call the ``start_session()`` method to instantiate a ``mongocxx::client_session``. 
Then, you can use either of the following APIs to perform a transaction:

- :ref:`Callback API <cpp-callback-api>`
- :ref:`Core API <cpp-core-api>`

.. _cpp-callback-api: 

Callback API 
~~~~~~~~~~~~

Use the callback API to start, execute, and commit a transaction with incorporated retry logic for ``TransientTransactionError`` and ``UnknownTransactionCommitResult`` errors. 
To implement this API, call the ``with_transaction()`` method on your ``mongocxx::client_session`` and pass in a 
callback specifying the sequence of operations you want to run in your transaction. 

The following code shows how to use the callback API to perform a transaction that inserts documents into collections in the ``sample_mflix`` database. 
When run, the code performs the following actions: 

1. Starts a session from the client using the ``start_session()`` method
#. Defines a callback function that specifies the operations to perform during the transaction
#. Creates an option object to set the write preference for the transaction operations

> #. Calls the ``with_transactions()`` method to manage the transaction, passing the callback and option object as arguments

Author: Hm I think this is probably ok. If you look at the code examples on other pages (e.g. the [GridFS](https://www.mongodb.com/docs/languages/cpp/cpp-driver/current/write/gridfs/#example) page), we sometimes add bulleted or numbered descriptions above the code example.

         instance as an argument to set options. For a full list of options, see `mongocxx::options::transaction <{+api+}/classmongocxx_1_1options_1_1transaction.html>`__
         in the API documentation.
       |
       | Raises an exception if the options are misconfigured, if there are network or other transient failures, or if there 
         are other errors such as a session with a transaction already in progress. If an error is returned with the ``TransientTransactionError`` label, 
         you can end the transaction and then retry it with the expectation that it will succeed. 
       |
       | To learn more about this method, see the :manual:`startTransaction()
         </reference/method/Session.startTransaction/>` guide in the {+mdb-server+} manual.

   * - ``commit_transaction()``
     - | Commits the active transaction on the current client session.
       |
       | Raises an exception if options are misconfigured, if there are network or other transient failures,
         or if there are other errors such as a session with no transaction in progress. If an error is returned with the ``UnknownTransactionCommitResult`` label,

>          you can end the transaction and then retry it with the expectation that it will succeed when the committed transaction satisfies the set write concern. 

Reviewer: Q: what does "committed transaction [satisfying] the set write concern" look like? Can't quite wrap my head around it

Author: I adapted this line from the [Rust Transactions page](https://www.mongodb.com/docs/drivers/rust/v2.8/fundamentals/transactions/). I think the write concern states the number of replica set members that the transaction operations should write to when it performs a write operation (e.g. write to the "majority" of replica set members). So the UnknownTransactionCommitResult error indicates that the driver doesn't know whether or not the write operations have been replicated across the preferred number of replica set members. 

So the logic here is that the active transaction will be re-committed until it can be certain that it has been replicated across enough sets to "satisfy" the set write concern (e.g. replicated across the majority).

       |
       | To learn more about this method, see the :manual:`commitTransaction()
         </reference/method/Session.commitTransaction/>` guide in the {+mdb-server+} manual. 

>    * - ``abort_transcaction()``

Reviewer: I:
```suggestion
   * - ``abort_transaction()``
```

     - | Ends the active transaction on the current client session. 
       |
       | Raises an exception if the options are misconfigured or if there are other errors such as 
         a session with no transaction in progress.
       |
       | To learn more about this method, see the :manual:`abortTransaction()
         </reference/method/Session.abortTransaction/>` guide in the {+mdb-server+} manual.

.. tip:: 
   
   The ``mongocxx::client_session`` class also provides methods to retrieve and modify session properties.
   To learn more, see `mongocxx::client_session <{+api+}/classmongocxx_1_1v__noabi_1_1client.html>`__ in the API documentation. 

> The following code shows how to use the core API to perform a transaction that inserts documents into collections in the ``sample_mflix`` database. 

Reviewer: ```suggestion
The following code shows how to use the core API to perform a transaction that inserts documents into the ``movies`` and ``comments`` collections in the ``sample_mflix`` database. 
```

When run, the code performs the following actions: 

1. Starts a session from the client using the ``start_session()`` method.
#. Creates an option object to set the write preference for the transaction operations. 
#. Calls the ``start_transaction()`` method to start a transaction, passing in the option object as an argument.
#. Attempts operations to insert documents into collections in the ``sample_mflix`` database. 
   If an operation encounters an error, the transaction is aborted and
   then retried if the error has the label ``TransientTransactionError``.  

> #. Commits the active transaction using the ``commit_transaction()`` method. If the commit encounters an error with the label ``UnknownTransactionCommitResult``, the commit is retried. 

Reviewer: S: Same as above - consider moving after the code?

Author: Same response as above

We recommend that you reuse your client for multiple sessions and transactions instead of
instantiating a new client each time.

.. warning::

   Use a ``mongocxx::client_session`` only with the ``mongocxx::client`` that created it. 
   Using a ``client_session`` with a different ``client`` results in operation errors.

.. important::
   
   Instances of ``mongocxx::client`` are not thread-safe. 
   Each ``mongoxcc::client`` instance and its child instances, including ``mongocxx::client_session``, should be used by a single thread at a time. 
   To learn more, see the :ref:`Thread and Fork Safety <cpp-thread-safety>` guide. 

.. _cpp-transaction-apis: 

Transaction APIs
----------------

The {+driver-long+} provides a callback API and a core API to manage the transaction lifestyle. 
Before you begin a transaction, you must call the ``start_session()`` method to instantiate a ``mongocxx::client_session``. 
Then, you can use either of the following APIs to perform a transaction:

- :ref:`Callback API <cpp-callback-api>`: High-level API that manages the life cycle of the transaction and automatically incorporates error handling logic.

> - :ref:`Core API <cpp-core-api>`: Low-level API that allows you to manually manage the life cycle of the transaction and implement custom error handling logic. 

Reviewer: ```suggestion
- :ref:`Core API <cpp-core-api>`: Low-level API that allows you to manage the life cycle of the transaction and implement custom error handling logic. 
```

If an error has the label ``TransientTransactionError`` and ``UnknownTransactionCommitResult``, the ``with_transaction()`` method automatically incorporates retry logic. 

The following code shows how to use the callback API to perform a transaction that inserts documents into the ``movies`` and ``comments`` collections in the ``sample_mflix`` database. 
This code performs the following actions: 

1. Starts a session from the client using the ``start_session()`` method.
#. Defines a callback function that specifies the operations to perform during the transaction.
#. Creates an option object to set the write concern for the transaction operations.
   To learn more about read and write semantics, see the :manual:`Read Concern/Write Concern/Read Preference </core/transactions/#read-concern-write-concern-read-preference>`
   section in the {+mdb-server+} manual. 
#. Calls the ``with_transaction()`` method to manage the transaction, passing the callback function and option object as arguments.

.. literalinclude:: /includes/write/transactions.cpp
   :language: cpp
   :dedent:
   :start-after: start-callback-api
   :end-before: end-callback-api

.. _cpp-core-api: 

Core API 
~~~~~~~~

> Use the core API to manually manage the life cycle of your transaction. The core API doesn't incorporate error handling logic 
> for ``TransientTransactionError`` and ``UnknownTransactionCommitResult`` errors, and instead provides the flexibility to implement custom error handling logic.
> 
> The core API requires explicit calls to start, commit, and abort the transaction using methods provided by the ``mongocxx::client_session`` interface.
> The following table describes these methods:  

Reviewer: ```suggestion
The core API requires explicit calls to start, commit, and abort the transaction using methods provided by the ``mongocxx::client_session`` interface. Use the core API to manually control the life cycle of your transaction. The core API doesn't incorporate error handling logic 
for ``TransientTransactionError`` and ``UnknownTransactionCommitResult`` errors, and instead provides the flexibility to implement custom error handling logic.

The following table describes the methods provided by the core API:  
```
Similar to what you did for the Callback API initial paragraphs
